### DISCUSSION TITLE
Get the solar generated energy stats

### DISCUSSION THREAD

**Discussion by shanrahul** [jblance/mpp-solar#176] [Q&A] [answered]
Hi , Is there any way to get the solar generated energy by day, month and year?

**Comment by jblance** [comment]
some inverters have commands for this information - what inverter model are you using
this code is basically just for querying the inverter - it doesnt maintain state my itself. If the inverter doesnt have this information you can use influx and grafana to store, aggregate and display data over time

**Reply by shanrahul** [reply]
I'm using [Easun](https://www.easunpower.com/products/EASUN-POWER-Bluetooth-Inverter-3000W-500Vdc-PV-230Vac-24Vdc-80A-MPPT-Solar-Charger-Support-Mobile-Monitoring-USB-LCD-Control) SMV3kw  inverter and I had an option to see the generated energy stats in the display of the inverter. My inverter is using the PI30 protocol and i refer the document you attached here , "PI30_PIP-GK_MK-Protocol.pdf" , there it says like this,

> 1.15 QED<YYYYMMDDnnn><cr>: Inquiry total energy in the day 
> Computer: QED<YYYYMMDDnnn><cr> 
> <YYYYMMDD> Y is an Integer number 0 to 9, M is an Integer number 0 to 9, D is an 
> Integer number 0 to 9. nnn is the checksum for QEDYYYYMMDD and the decimal number 
> towards lower 8 digits. 
> Device: (NNNNNN<cr>. 
> N is an Integer number 0 to 9. The unit is Wh. 
> For example: 
> Computer：QED20110701102<cr> 
> Device: (030123<cr> 
> Means: in 2010/09/26, the total energy is: 30123Wh
> 

So is there any way to use this in the code ?

**Comment by jblance** [comment, marked as answer]
You could try with the PI30MAX protocol for the QED command

**Comment by smrruby** [comment]
I'm using the "ET" query under Protocol 18 for this on an LVX6048 which provides a value that should be an accumulator for total energy generated
```
ET                              Query total generated energy -- queries total generated energy from the Inverter
```

However, while the number generally increases over time (as a meter/accumulator) should, sometimes it randomly shrinks between readings for reasons I can't explain.  I don't believe the issue is inside mpp-solar because I would see sudden decreases in the value using the app as well.  

Anybody else experience these inverters suddenly reducing the metered energy generation?  Find any fixes (settings, etc?)?

this value seems reasonable until it randomly decreases over 100, or even over 1000 KWH between reads
```
2023-01-22 14:29:19,462:DEBUG:hidrawio:send_and_receive@36: 1 chunk send
2023-01-22 14:29:20,382:DEBUG:hidrawio:send_and_receive@69: usb response was: b"^D01100370583'\xce\r"
```
```
Command: ET - Query total generated energy
------------------------------------------------------------
Parameter                       Value           Unit
generated_energy                370583          kWh
```

**Reply by wolf76man** [reply]
Hi!
I am using P16 protocol and i see missing commands: QEM, QEY and QEH. 
Is it posible to add? Can i do it on my local to test? How?

Thanks for your work! This is very usuefull to me!

**Reply by jblance** [reply]
created an issue for this #324 

if you know how to run from the source, making the changes involves editing the protocol definition for [pi16](https://github.com/jblance/mpp-solar/tree/master/mppsolar/protocols/pi16.py)

**Reply by wolf76man** [reply]
Thanks!
I am a developer. But never touched python before.
I'll give it a go this week. 
Thanks again

**Reply by jblance** [reply]
there is a QED definition that can be used as an example

**Comment by wolf76man** [comment]
Yes. I saw it. It should be almost the same for all.
I am fluent in java. But i couldn't find a way to read the inverter USB. I have code in x86 that work ok. But on arm is not working.. 
Ill set a python env to do some test before attempt this. I want to understand how is the best way to go.
Thank John

**Comment by jblance** [comment]
if you wait a few days I will make the changes
otherwise, a separate python venv is best start

**Reply by jblance** [reply]
a quick attempt at adding these in latest source

**Reply by wolf76man** [reply]
thanks. ill check
i tried to run locally. i found error in requirements.txt:
mpp-solar should be mppsolar. no?

now i have error ModuleNotFoundError: No module named 'setuptools'
not sure why its all installed.
i am a rookie hahaha

**Reply by jblance** [reply]
yeah installing from source is always a painful "oh i need another package" process

if you use a venv, you end up with a isolated environment, so things need to be installed there (and sudo breaks out of the venv)

**Reply by wolf76man** [reply]
Hi!
Sudo was creating issues. thanks for the tip.
i was able to run some commands from my local after installing locally with:
pip install -e .

but if i try to run others i get
 pkg_resources.DistributionNotFound: The 'psycopg2' distribution was not found and is required by mppsolar

this is familiar to you?
Thanks

**Reply by wolf76man** [reply]
Made it work with poetry.
If i add something to the code ill share with you.
Thanks.

**Comment by jblance** [comment]
Great to hear. Just documenting what you've done is a great help (docs
always get left behind)
